CTRE CANdles

Bright-Shard opened the pull request:
Merge support for CTRE CANdles into master, and rename the `talon_interface` package to `ctre_interfaces`. All code built on Ben's laptop, and the CANdles worked on the robot-in-a-box.

kjaget commented on scripts/deploy.sh:
```diff
@@ -8,7 +8,7 @@ ROBORIO_ADDR=10.9.0.2
 
 # This can be an array of IP address if there are multiple Jetsons
 #JETSON_ADDR=( )
-JETSON_ADDR=(10.9.0.8 10.9.0.9)
```
Revert this so robot deploys don't stop working once we merge

Bright-Shard replied:
This is from before I reverted deploy.sh, resolving

kjaget commented on zebROS_ws/src/candle_controller/src/candle_controller.cpp:
```diff
@@ -0,0 +1,255 @@
+        const Brightness brightness = *(this->brightness_buffer.readFromRT());
+
+        // Brightness isn't exclusive to colours/animations, so it gets special treatment
+        if (brightness.time > this->last_write) {
```
It might be OK to rewrite brightness each time through as long as the lower level command code does an early out if new_brightness == current_brightness.  That would simplify the controller code a good bit.

Bright-Shard replied:
I think the main reason I had the if statement was for the debug message. Once I'm confident this all works, we can remove that and just write brightness each time.

```c++
if (brightness.time > this->last_write) {
    ROS_INFO_STREAM("Writing new brightness to CANdle");
    this->candle_handle->setBrightness(brightness.brightness);
}
```

kjaget commented on zebROS_ws/src/candle_controller/src/candle_controller.cpp:
```diff
@@ -0,0 +1,255 @@
+            this->led_buffer.writeFromNonRT(leds);
+
+            return true;
+        } else {
```
Stylistic option nitpicking - if code returns from an if block, the else can be removed and the code in the else block made unconditional.

I have no strong opinions which is best, doesn't matter functionally.

Bright-Shard replied:
I'm going to leave this as-is - I find it easier to read the `else` block than assuming it from the return statement.

kjaget commented on zebROS_ws/src/ctre_interfaces/include/ctre_interfaces/candle_command_interface.h:
```diff
@@ -0,0 +1,83 @@
+#include <vector>
+#include <optional>
+
+using namespace std;
```
Don't use `using namespace` inside header files - https://stackoverflow.com/questions/14575799/using-namespace-std-in-a-header-file

Bright-Shard replied:
wasn't aware of this issue, should be fixed now

kjaget replied:
Yeah, one of many fun c++ pitfalls.

kjaget commented on zebROS_ws/src/ctre_interfaces/include/ctre_interfaces/candle_state_interface.h:
```diff
@@ -0,0 +1,133 @@
+#include <hardware_interface/internal/hardware_resource_manager.h>
+#include "state_handle/state_handle.h"
+
+using namespace std;
```
Another using namespace std to remove

Bright-Shard commented:
I've reverted the `deploy.sh` script, but I also added a `deploy_robot_in_a_box.sh` script that will deploy to the robot in a box (it doesn't have a second jetson, so normal deploys don't work).

kjaget commented on zebROS_ws/src/ctre_interfaces/src/candle_state_interface.cpp:
```diff
@@ -0,0 +1,128 @@
+}
+
+void CANdleHWState::setLED(int id, CANdleColour led) {
+    this->leds[id] = led;
```
What happens if code tries to set an LED on the strip (ids 9 and above)?

Bright-Shard replied:
Tested this on the robot-in-a-box, I got animations and normal colours working on a neopixel strip connected to the CANdle. I'm assuming that's what you're talking about so I'll mark this as resolved

kjaget replied:
I was more concerned with the potential for writing past the end of the size of leds, not sure how resizing works for that vector

Bright-Shard replied:
well, it works regardless, lol. I can write to those LEDs

kjaget commented on zebROS_ws/src/ros_control_boilerplate/src/candle_convert.cpp:
```diff
@@ -0,0 +1,133 @@
+            return ColorFlowAnimation::Direction::Forward;
+        case 1:
+            return ColorFlowAnimation::Direction::Backward;
+    }
```
Normally I like to thrown in a default case which prints an error. It very rarely pops up once code is working but does help catch weird cases during debugging.

Bright-Shard replied:
Added a default case

kjaget commented on zebROS_ws/src/ros_control_boilerplate/src/frc_robot_interface.cpp:
```diff
@@ -2570,6 +2585,139 @@ void FRCRobotInterface::write(const ros::Time& time, const ros::Duration& period
+		}
+
+		// Enabled
+		//ROS_INFO_STREAM("Candle_status_3");
```
If code looks good, probably best to remove the commented-out printouts.

Bright-Shard replied:
Yup, completely forgot about those... removed now

kjaget commented on zebROS_ws/src/ros_control_boilerplate/src/frc_robot_interface.cpp:
```diff
@@ -2570,6 +2585,139 @@ void FRCRobotInterface::write(const ros::Time& time, const ros::Duration& period
+			ROS_INFO_STREAM("Candle_status_4.1");
+			// Convert from CANdleAnimation to the appropriate CTRE animation class
+			if (candle_animation.class_type == CANdleAnimationClass::BaseStandard) {
+				ROS_INFO_STREAM("Candle_status_4.1.1");
```
And definitely remove these non-printed out debugging bits
Keep any actual status printouts when things change, though - those are nice for confirming things work, just as long as they're not printing every loop iteration

kjaget commented on zebROS_ws/src/ros_control_boilerplate/src/frc_robot_interface.cpp:
```diff
@@ -2570,6 +2585,139 @@ void FRCRobotInterface::write(const ros::Time& time, const ros::Duration& period
+				ROS_INFO_STREAM("Candle_status_4.1.1");
+				BaseStandardAnimation base_animation = candle_convert::convertBaseStandardAnimation(candle_animation);
+				ROS_INFO_STREAM("Candle_status_4.1.2");
+				animation = &base_animation;
```
This isn't going to work, or at least not consistently. animation is going to point to base_animation, and base_animation goes out of scope and becomes invalid once leaving the block it is declared in.  
You might get lucky for a while and that memory not be overwritten by something else, but it will break at some point.

An idea for fixing this. Make a 
`std::shared_ptr<Animation>` that's a member var of frc robot interface.
Make convertAnimation return a `shared_ptr<BaseAnimation>`, constructed using something like `std::make_shared<SpecificAnimationType>(arg, arg, arg);`
Assign the results of convertAnimation to the new class member var.
Same with convertBaseTwoAnimation().

Making the shared_ptr holding the animation a member var will make sure it stays in scope. Although looking at the docs, it just has to stay in scope until Animate() is called, so it could also just be a local declared before the if statement.  That is, just make the current *animation var a shared_ptr and proceed like above.

kjaget replied:
Why this works - make_shared will dynamically allocate memory for the object, and that memory will only be deallocated when the last shared_ptr() pointing to it goes out of scope.

kjaget replied:
Also, you might be able to use std::unique_ptr / std::make_unique instead, but that might get weird with returning it from a function.

Bright-Shard replied:
I applied the shared_ptr fix (and it builds). I might experiment with unique pointers tomorrow, but practice has ended today.

kjaget commented on zebROS_ws/src/ros_control_boilerplate/src/frc_robot_interface.cpp:
```diff
@@ -2570,6 +2585,139 @@ void FRCRobotInterface::write(const ros::Time& time, const ros::Duration& period
+		// LEDs
+		//ROS_INFO_STREAM("Candle_status_5");
+		vector<LEDGroup> led_groups;
+		if (candle_command.ledGroupChanged(led_groups)) {
```
Think about the possibilities if both animations and led groups are used.  Should there be a call to disable animations if led groups are set?  Should there be an explicit clear / stop animation in the command interface?

kjaget replied:
Likewise, should setting an animation first call setLED (0,0,0) on all the LEDs in range, or does this happen automatically?

kjaget reviewed (changes requested):
A few major things to look at along with some minor stuff.  
Excited to see this running on the robot (in a box), lots of fun potential to work with.

kjaget commented:
> I've reverted the `deploy.sh` script, but I also added a `deploy_robot_in_a_box.sh` script that will deploy to the robot in a box (it doesn't have a second jetson, so normal deploys don't work).

Could you make it a command line option for deploy.sh? deploy gets updates more frequently than it should, would rather not have two copies floating around
You could make an alias to deploy with that flag in https://github.com/FRC900/2022RobotCode/blob/master/zebROS_ws/command_aliases.sh to create a deploy_robot_in_a_box command

Bright-Shard commented:
> > I've reverted the `deploy.sh` script, but I also added a `deploy_robot_in_a_box.sh` script that will deploy to the robot in a box (it doesn't have a second jetson, so normal deploys don't work).
> 
> Could you make it a command line option for deploy.sh? deploy gets updates more frequently than it should, would rather not have two copies floating around You could make an alias to deploy with that flag in https://github.com/FRC900/2022RobotCode/blob/master/zebROS_ws/command_aliases.sh to create a deploy_robot_in_a_box command

I modified deploy.sh directly - adding the "box" argument will deploy to the robot in a box instead of the actual robot. Doing some testing now and will push afterwards.
```c++
# This can be an array of IP address if there are multiple Jetsons
#JETSON_ADDR=( )
# If we're deploying to the robot in a box, only deploy to 1 jetson (it doesn't have 2)
if [ $1 == "box" ]; then
    JETSON_ADDR=(10.9.0.8)
else
    # On the actual robot, deploy to both jetsons
    JETSON_ADDR=(10.9.0.8 10.9.0.9)
fi
```

kjaget commented on scripts/deploy.sh:
```diff
@@ -8,7 +8,13 @@ ROBORIO_ADDR=10.9.0.2
 # This can be an array of IP address if there are multiple Jetsons
 #JETSON_ADDR=( )
-JETSON_ADDR=(10.9.0.8 10.9.0.9)
+# If we're deploying to the robot in a box, only deploy to 1 jetson (it doesn't have 2)
```
This should probably be integrated into the command line arg parsing starting around line 40.

kjaget commented:
Moving over to 2023 repo, closing
